test/diff/line.js

```diff
@@ -138,16 +138,39 @@ describe('diff/line', function() {
     it('should ignore leading and trailing whitespace', function() {
       const diffResult = diffTrimmedLines(
         'line\nvalue \nline',
-        'line\nvalue\nline');
-      expect(convertChangesToXML(diffResult)).to.equal('line\nvalue\nline');
+        'line \nvalue\nline');
+      expect(convertChangesToXML(diffResult)).to.equal('line \nvalue\nline');
+    });
+
+    it('should not consider adding whitespace to an empty line an insertion', function() {
+      const diffResult = diffTrimmedLines('foo\n\nbar', 'foo\n \nbar');
+      expect(convertChangesToXML(diffResult)).to.equal('foo\n \nbar');
     });
 
     it('should handle windows line endings', function() {
       const diffResult = diffTrimmedLines(
         'line\r\nold value \r\nline',
         'line\r\nnew value\r\nline');
-      expect(convertChangesToXML(diffResult)).to.equal('line\r\n<del>old value\r\n</del><ins>new value\r\n</ins>line');
+      expect(convertChangesToXML(diffResult)).to.equal('line\r\n<del>old value \r\n</del><ins>new value\r\n</ins>line');
     });
+
+    it('should be compatible with newlineIsToken', function() {
+      const diffResult = diffTrimmedLines(
+        'line1\nline2\n   \nline4\n \n',
+        'line1\nline2\n\n\nline4\n   \n',
+        {newlineIsToken: true}
+      );
+      expect(convertChangesToXML(diffResult)).to.equal('line1\nline2\n<del>   </del>\n<ins>\n</ins>line4\n   \n');
```

Comment:
Hmmm... this made sense to me when I wrote it, but on reflection it seems wrong for `<del>   </del>` to show up here. After all, a completely empty line and a line with a load of spaces are meant to be equal under `ignoreWhitespace`, right?

Reply:
On even further reflection: I think the problem here is that:
* for `newlineIsToken` mode to make sense at all, completely blank line content shouldn't be a token (otherwise it's *strictly* the case that every second token is a newline character, so the newline characters will *always* be preserved, and the intended purpose of `newlineIsToken` - to allow changes to newline structure to be reflected in diff results independent of line content changes - is undermined).
* but to ensure that joining the non-removed bits of the diff actually yields the new text, non-blank, whitespace-only lines DO need to get a token!

Combining the two requirements above, we end up forced to treat deleting all the whitespace from a whitespace-only line as a deletion in `newlineIsToken` + `ignoreWhitespace` mode. I think this is therefore the right thing to do (if we're going to allow those settings to be used together at all) and I should just note it as a curiosity in the docs.